source/cipher.c

```diff
@@ -88,6 +88,62 @@ static enum aws_cryptosdk_sha_version aws_cryptosdk_which_sha(enum aws_cryptosdk
     }
 }
 
+static bool str_eq(const char *const str1, const char *const str2) {
+    if ((str1 && !str2) || (!str1 && str2)) {
+        return false;
+    }
+    if (!str1 && !str2) {
+        return true;
+    }
+#ifdef CBMC
+    /* TODO: Bound the alg_props_string size to 128 characters (or any
+       reasonable max size).  This is only useful for CBMC to work,
+       not a real precondition. (We have to include that in the
+       validity check for alg_props (checking for that bound) so that
+       it is at least checked in debug_mode. */
+    /* For now just assert that they both point somewhere */
+    return true;
+#else
+    return strcmp(str1, str2) == 0;
+#endif
+}
+
+static bool alg_properties_equal(const struct aws_cryptosdk_alg_properties alg_props1,
+                                 const struct aws_cryptosdk_alg_properties alg_props2) {
+    return
+        /* Note: strcmp messes with CBMC so we should either define
+           this to be different for CBMC or bound the size of names */ 
+        /* strcmp(alg_props1.md_name, alg_props2.md_name) == 0 && */
+        /* strcmp(alg_props1.cipher_name, alg_props2.cipher_name) == 0 && */
+        /* strcmp(alg_props1.alg_name, alg_props2.alg_name) == 0 && */
+        
+        /* Note: for now we don't check the equiality of the alg_impl
+         * struct. Is there a way to check for equality of the impl
+         * structure? */
+        alg_props1.data_key_len == alg_props2.data_key_len &&
+        alg_props1.content_key_len == alg_props2.content_key_len &&
+        alg_props1.iv_len == alg_props2.iv_len &&
+        alg_props1.tag_len == alg_props2.tag_len &&
+        alg_props1.signature_len == alg_props2.signature_len &&
+        alg_props1.alg_id == alg_props2.alg_id;
+}
+
+bool aws_cryptosdk_alg_properties_is_valid(const struct aws_cryptosdk_alg_properties *const alg_props) {
+    if (alg_props == NULL) {
+        return false;
+    }   
+    enum aws_cryptosdk_alg_id id = alg_props->alg_id;
+    struct aws_cryptosdk_alg_properties *std_alg_props = aws_cryptosdk_alg_props(id);
```

Comment:
Is it reasonable to expect that the `alg_props` structure will always be initialized as shown in `aws_cryptosdk_alg_props`? Can this function act as the specification of a valid `alg_props` structure as shown here?

Reply:
`aws_cryptosdk_alg_prop` looks up and returns the algorithm properties for a particular algorithm id and is currently used to initialize `alg_props` within the SDK. Refer: [cipher.h#L102](https://github.com/aws/aws-encryption-sdk-c/blob/master/include/aws/cryptosdk/cipher.h#L102).  However,  one could also do a static initialization like we have done here: [cipher.c#L40](https://github.com/aws/aws-encryption-sdk-c/blob/master/source/cipher.c#L40)